libraries/AP_Follow/AP_Follow.cpp

```diff
@@ -0,0 +1,336 @@
+    Location target_loc;
+    Vector3f veh_vel;
+    if (!get_target_location_and_velocity(target_loc, veh_vel)) {
+        return false;
+    }
+
+    // calculate difference
+    const Vector3f dist_vec = location_3d_diff_NED(current_loc, target_loc);
+
+    // fail if too far
+    if (is_positive(_dist_max.get()) && (dist_vec.length() > _dist_max)) {
+        return false;
+    }
+
+    // initialise offsets from distance vector if required
+    init_offsets_if_required(dist_vec);
+
+    // get offsets
+    Vector3f offsets;
+    if (!get_offsets_ned(offsets)) {
+        return false;
+    }
+
+    // return results
+    dist = dist_vec;
+    dist_with_offs = dist_vec + offsets;
+    vel = veh_vel;
+    return true;
+}
+
+// get target's heading in degrees (0 = north, 90 = east)
+bool AP_Follow::get_target_heading(float &heading) const
+{
+    // exit immediately if not enabled
+    if (!_enabled) {
+        return false;
+    }
+
+    // check for timeout
+    if ((_last_heading_update_ms == 0) || (AP_HAL::millis() - _last_heading_update_ms > AP_FOLLOW_TIMEOUT_MS)) {
+        return false;
+    }
+
+    // return latest heading estimate
+    heading = _target_heading;
+    return true;
+}
+
+// handle mavlink DISTANCE_SENSOR messages
+void AP_Follow::handle_msg(const mavlink_message_t &msg)
+{
+    // exit immediately if not enabled
+    if (!_enabled) {
+        return;
+    }
+
+    // skip our own messages
+    if (msg.sysid == mavlink_system.sysid) {
+        return;
+    }
+
+    // skip message if not from our target
+    if ((_sysid_to_follow != 0) && (msg.sysid != _sysid_to_follow)) {
+        if (_sysid == 0) {
+            // maybe timeout who we were following...
+            if ((_last_location_update_ms == 0) || (AP_HAL::millis() - _last_location_update_ms > AP_FOLLOW_TIMEOUT_MS)) {
+                _sysid_to_follow = 0;
+            }
+        }
+        return;
+    }
+
+    // decode global-position-int message
+    if (msg.msgid == MAVLINK_MSG_ID_GLOBAL_POSITION_INT) {
+
+        uint32_t now = AP_HAL::millis();
+
+        // get estimated location and velocity (for logging)
+        Location loc_estimate{};
+        Vector3f vel_estimate;
+        UNUSED_RESULT(get_target_location_and_velocity(loc_estimate, vel_estimate));
+
+        // decode message
+        mavlink_global_position_int_t packet;
+        mavlink_msg_global_position_int_decode(&msg, &packet);
+
+        // ignore message if lat and lon are zero
+        if ((packet.lat != 0) || (packet.lon != 0)) {
+            return;
+        }
+
+        _target_location.lat = packet.lat;
+        _target_location.lng = packet.lon;
+        _target_location.alt = packet.alt / 10;     // convert millimeters to cm
+        _target_velocity_ned.x = packet.vx * 0.01f; // velocity north
+        _target_velocity_ned.y = packet.vy * 0.01f; // velocity east
+        _target_velocity_ned.z = packet.vz * 0.01f; // velocity down
+        _last_location_update_ms = now;
```

Comment:
It would be a LOT better if we used the source timestamp of the message instead of now :)

Reply:
Yes, although I think we need to store both times.  At the moment _last_location_update_ms is used for two purposes:

1. timeout calcs if we haven't received a message in a while
2. calculating the dt to extrapolate the current position from the last update message

the timeout should use our system time while the dt could use the system time of the other vehicle received in the message.

I think I'll add this as an item on the to-do list so we don't forget about it.